### PR TITLE
Function mapped methods refactor

### PR DESCRIPTION
**Pull request by Kevin-TD**

### CODE REVIEW

**Review by kelloggm — CHANGES_REQUESTED** [review]

File: include/CalledMethods.h
Comment (Kevin-TD) [reply include/CalledMethods.h:9]: Fixing this to use MethodsSet instead

File: include/ProgramRepresentation/FullProgram.h
Comment (Kevin-TD) [reply include/ProgramRepresentation/FullProgram.h:6]: Yeah it's for a single file; updating the comment accordingly

File: include/ProgramRepresentation/FullProgram.h
Comment (Kevin-TD) [reply include/ProgramRepresentation/FullProgram.h:12]: I'll name it to that instead; I was sticking with using the `name` field for these classes but yeah, it's better to make them more descriptive

File: src/DataflowPass.cpp
Comment (Kevin-TD) [reply src/DataflowPass.cpp:22]: Yeah they don't really have reason to be capitalized. It was probably the case that when I was writing it it was convenient or nice but I don't see any reason for it now

File: src/ProgramRepresentation/ProgramFunction.cpp
Comment (Kevin-TD) [reply src/ProgramRepresentation/ProgramFunction.cpp:31]: I now realize that it should error instead because it returning a `new ProgramPoint` really shouldn't be happening. For now, I'll add a todo saying that it should error instead. Also I'll update the docs

Comment (kelloggm) [reply src/ProgramRepresentation/ProgramFunction.cpp]: Why add a TODO and not just have it error now? I don't think it would be that hard to change this to throw some kind of exception if this case is really expected never to occur.

File: src/ProgramRepresentation/MethodsSet.cpp
Comment (kelloggm) [inline src/ProgramRepresentation/MethodsSet.cpp:12]: This copy semantics should be documented, both here and for the getter. It's somewhat unintuitive that `MethodsSet::getMethods` doesn't actually return the backing set, but instead returns a copy.

File: src/ProgramRepresentation/ProgramFunction.cpp
Comment (kelloggm) [inline src/ProgramRepresentation/ProgramFunction.cpp:51]: Stopping the program is relatively easy, so let's just do this rather than leaving a TODO and failing silently.

**Review by kelloggm — CHANGES_REQUESTED** [review]

File: src/ProgramRepresentation/MethodsSet.cpp
Comment (Kevin-TD) [reply src/ProgramRepresentation/MethodsSet.cpp:12]: I initially did this to be 100% sure a copy was being made since I initially suspected that I might've been returning a reference and that was causing seg faults with the lub function. But now I realize I don't need them, so they'll be removed. The issue had something to do with `std::set_intersection` and `std::set_union.` When I wrote the code for lub like this: 
```
void CalledMethods::leastUpperBound(MethodsSet &preMethods,
                                    MethodsSet &curMethods,
                                    MethodsSet &result) {
  std::set<std::string> res;

  std::set_intersection(preMethods.getMethods().begin(), preSet.getMethods().end(), curSet.getMethods().begin(),
                        curSet.getMethods().end(), std::inserter(res, res.begin()));
  result.setMethods(res);
}
```
I received a seg fault, but changing it to: 
```
void CalledMethods::leastUpperBound(MethodsSet &preMethods,
                                    MethodsSet &curMethods,
                                    MethodsSet &result) {
  std::set<std::string> res;
  std::set<std::string> preSet = preMethods.getMethods();
  std::set<std::string> curSet = curMethods.getMethods();

  std::set_intersection(preSet.begin(), preSet.end(), curSet.begin(),
                        curSet.end(), std::inserter(res, res.begin()));
  result.setMethods(res);
}
```
did not cause me any seg faults. So as far as I can tell, the issue dealt with `set_intersection`/`set_union`, and I didn't find any reference/memory issues elsewhere. But I'm not sure why one version causes a seg fault and the other doesn't, as they should effectively be doing the same thing.

Comment (kelloggm) [reply src/ProgramRepresentation/MethodsSet.cpp]: Yes, I agree this is really surprising. Can you add a TODO here with a link to your comment saying that we should investigate this at some point - it makes me uncomfortable, but if it's working now I'm okay with moving on.

**Review by kelloggm — APPROVED** [review]
LGTM, modulo adding a single TODO. I don't need to review this again before it is merged.

File: src/ProgramRepresentation/MethodsSet.cpp
Comment (Kevin-TD) [reply src/ProgramRepresentation/MethodsSet.cpp]: After some research, I realized why the first one seg faults but the second doesn't. It deals with the `.begin` and `.end` iterators.
When I make a call to `getMethods`, it returns a copy of some `set<string>`. In the first one, I'm making two different copies of the `set<string>` by calling std::set_intersection(preMethods.**getMethods**().begin(), preSet.**getMethods**().end() ..., and the beginning and end iterators for these sets likely point to unrelated spots in memory, so it's likely for the program to try to reach inaccessible memory, causing a seg fault. 
For the second one, I'm only making one copy for `preSet` and one copy for `curSet`, so when I call `.begin` and `.end`, the iterators are correct and there won't be a bad memory access.

Comment (kelloggm) [reply src/ProgramRepresentation/MethodsSet.cpp]: That is such a classic C++ [footgun](https://news.ycombinator.com/item?id=17393292)! Thanks for figuring it out.

**Pull request merged**
